cctp_primitives/src/proving_system/verifier/mod.rs

```diff
@@ -0,0 +1,126 @@
+use algebra::{
+    SerializationError, SemanticallyValid,
+};
+use crate::{
+    type_mapping::*,
+    proving_system::error::ProvingSystemError,
+    utils::proof_system::ProvingSystemUtils,
+};
+use rand::RngCore;
+use crate::utils::serialization::deserialize_from_buffer;
+
+pub mod certificate;
+// To be defined
+//pub mod ceased_sidechain_withdrawal;
+pub mod batch_verifier;
+
+/// Utility enum, allowing the cryptolibs to pass data and
+/// specify the proving system type at the same type.
+#[derive(Clone)]
+pub enum RawVerifierData {
+    CoboundaryMarlin{ proof: Vec<u8>, vk: Vec<u8> },
+    Darlin{ proof: Vec<u8>, vk: Vec<u8> }
+}
```

Comment:
May be I'll ask something stupid. Is there any case when user inputs should be passed from other crypto lib as raw data too?

Reply:
For the cryptolibs we pass specific data to specific functions (like the CertificateProofUserInputs data structure) that build the actual public input of the circuit. Maybe we can, for generic usage, add a vector of bytes and deserialize field elements out of them, but it would create confusion cryptolibs side. Since we do not aim to be that much generic here, I would suggest to forget it for the moment.

Reply:
Ok